docs/scripts/check_imports.py

```diff
@@ -0,0 +1,129 @@
+import importlib
+import json
+import logging
+import os
+import re
+import warnings
+from pathlib import Path
+from typing import List, Tuple
+
+logger = logging.getLogger(__name__)
+
+DOCS_DIR = Path(os.path.abspath(__file__)).parents[1] / "docs"
+import_pattern = re.compile(
+    r"import\s+(\w+)|from\s+([\w\.]+)\s+import\s+((?:\w+(?:,\s*)?)+|\(.*?\))", re.DOTALL
+)
+
+
+def _get_imports_from_code_cell(code_lines: str) -> List[Tuple[str, str]]:
+    """Get (module, import) statements from a single code cell."""
+    import_statements = []
+    for line in code_lines:
+        line = line.strip()
+        if line.startswith("#") or not line:
+            continue
+        # Join lines that end with a backslash
+        if line.endswith("\\"):
+            line = line[:-1].rstrip() + " "
+            continue
+        matches = import_pattern.findall(line)
+        for match in matches:
+            if match[0]:  # simple import statement
+                import_statements.append((match[0], ""))
+            else:  # from ___ import statement
+                module, items = match[1], match[2]
+                items_list = items.replace(" ", "").split(",")
+                for item in items_list:
+                    import_statements.append((module, item))
+    return import_statements
+
+
+def _extract_import_statements(notebook_path: str) -> List[Tuple[str, str]]:
+    """Get (module, import) statements from a Jupyter notebook."""
+    with open(notebook_path, "r", encoding="utf-8") as file:
+        notebook = json.load(file)
+    code_cells = [cell for cell in notebook["cells"] if cell["cell_type"] == "code"]
+    import_statements = []
+    for cell in code_cells:
+        code_lines = cell["source"]
+        import_statements.extend(_get_imports_from_code_cell(code_lines))
+    return import_statements
+
+
+def _get_bad_imports(import_statements: List[Tuple[str, str]]) -> List[Tuple[str, str]]:
+    """Collect offending import statements."""
+    offending_imports = []
+    for module, item in import_statements:
+        try:
+            if item:
+                try:
+                    # submodule
+                    full_module_name = f"{module}.{item}"
+                    importlib.import_module(full_module_name)
+                except ModuleNotFoundError:
+                    # attribute
+                    try:
+                        imported_module = importlib.import_module(module)
+                        getattr(imported_module, item)
+                    except Exception as e:
+                        offending_imports.append((module, item))
+                except Exception as e:
```

Comment:
Why a broad exception?

Reply:
Added in specific exceptions I was expecting. These are broad in case I missed something among ImportError, AttributeError, or ModuleNotFoundError.